### DISCUSSION TITLE
General Issue: Clean-up Round 2

### DISCUSSION THREAD

**Discussion by bundfussr** [pharmaverse/admiral#2010] [Announcements]
### Background Information

To achieve a "stable" admiral version we need a clean-up of admiral, i.e.,
- avoid that functionality of functions overlaps
- name function and arguments consistently
- close gaps in functionality

This issue is intended to collect functions which should be revised and identify gaps.

@pharmaverse/admiral , please feel free to add.

**Comment by bundfussr** [comment]
From @kaz462 :

There are four functions for deriving extreme event/record as a new parameter, the function names are a bit confusing to me when I try to compare them: `_record(s)`/`_event`, with/without `_param`


  | Data source:      one or multiple | Extreme is ordered by - | filter/dataset source argument name
-- | -- | -- | -- 
1.[`derive_extreme_records`](https://pharmaverse.github.io/admiral/cran-release/reference/derive_extreme_records.html)| one | `order` argument | `filter`/ |
2.[`derive_extreme_event`](https://pharmaverse.github.io/admiral/cran-release/reference/derive_extreme_event.html)  | one | the order of `events` entered and the `order` argument |
3.[`derive_param_extreme_event`](https://pharmaverse.github.io/admiral/cran-release/reference/derive_param_extreme_event.html) | one  |`order` argument| `filter_source`/`dataset_source`
4.`derive_param_extreme_record` | multiple   |  `order` argument   | /`source_datasets` |



* What do you think of combining some functions? e.g., 

     * function 3 is a special case of function 2 

     * function 4 is an extension of function 2 by allowing multiple data sources  
     * function 1 and 3 has been combined through https://github.com/pharmaverse/admiral/pull/1882

* Should we make some argument name consisnent across those functions?

**Comment by bundfussr** [comment]
Should all flag functions provide `true_vlue` and `false_value` argument?

**Reply by zdz2101** [reply]
I actually don't think its necessary, feels like if they wanted to modify it from the usual "Y" values, its a really simple `dplyr::mutate` statement they can address on their own as a post-processing step, the more arguments we add to functions only adds more complexity to our codebase

**Comment by bundfussr** [comment]
`derive_summary_records()` should provide the `dataset_add` argument.
Consider the following use case.
```
adae <- adae %>%
  mutate(AVAL = 1)

adsaftte_aetot <- derive_summary_records(
  adae,
  filter = ANL01FL == "Y",
  set_values_to = exprs(
    PARAMCD = "AETOT1",
    PARAM = "Number of occurrences of any adverse event",
    PARCAT1 = "Total Occurrences",
    ANL01FL = "Y"
  ),    
  by_vars = exprs(STUDYID, USUBJID),
  analysis_var = AVAL,
  summary_fun = sum
) %>%
  filter(PARAMCD == "AETOT1") %>%
  select(STUDYID, USUBJID, PARAMCD, PARAM, PARCAT1, AVAL, ANL01FL)
```
The `filter()` and `select()` call at the end could be omitted if the `dataset_add` argument is available.

**Comment by bundfussr** [comment]
The arguments of `derive_vars_joined()`, `derive_var_joined_exist_flag()`, and `filter_joined()` should be made consistent. For example, all functions should provide the `dataset_add` argument.

**Comment by zdz2101** [comment]
@bms63 should we actually turn this into discussion with the respective individual pieces of consolidation as their own issues?

Link to mural of function arguments for viewing: https://app.mural.co/t/admiraldevelopmentteam6136/m/admiraldevelopmentteam6136/1689188235153/04c3276047791dc443bc22046e0804924a1e6137?sender=u2186aa07ff6a351a07a19987

**Comment by bms63** [comment]
Yes lets move to discussion - also makes it easier to follow threads

**Comment by zdz2101** [comment]
Many of our function arguments serve similar/same purposes but have different names, here are some recommended swaps, we can modify the table such that by the end/after all talk it becomes the "Definition of done" 

| argument | function | recommended  | implemented?| Strategy Addition/Update| All Done? |
|--------|-------|------|------|------|------|
| key_vars | consolidate_metadata | by_vars |
| check_vars | consolidate_metadata | check_type | Questioning |
| check_keys | consolidate_metadata | check_unqiue | | Needed |
| start_datetime | create_single_dose_dataset | start_date | Will Not| Needed |
| end_datetime | create_single_dose_dataset | end_date | Will Not| Needed |
| dataset_expected_obs | derive_expected_records | dataset_ref | #2037 | #306|
| dataset_expected_obs | derive_locf_records | dataset_ref |  #2037  | #306|
| dataset_adsl | derive_param_exist_flag | dataset_add | Already Done | #306 |
| dataset_adsl | derive_param_extreme_event | dataset_add | Being Deprecated | Needed |
| dataset_adsl | derive_param_tte | dataset_add | Most likely leave as is| #306 |
| na_val | derive_var_shift | missing_value | #2014 | #296|
|trt_start_date | derive_var_trtemfl | start_date | Will Not | Needed |
|trt_end_date | derive_var_trtemfl| end_date | Will Not | Needed |
| dataset_facm | derive_vars_atc | dataset_add | Will Not| Needed |
| reference_date | derive_vars_dy | start_date | #2027 |
| missing_values | derive_vars_joined | missing_value | Leave as is | #296|
| missing_values | derive_vars_merged | missing_value | Leave as is |  #296|
| dataset_merge | derive_vars_transposed | dataset_add |  Needed |
| cond | max_cond | condition |
| cond | min_cond | condition |
| wt | compute_egfr | weight | #2020|
| .../args | lots of places | sources | #2023 |

**Reply by bundfussr** [reply]
We can not rename `trt_start_date` and `trt_end_date` in `derive_var_trtemfl()` because `start_date` and `end_date` are already arguments of the function.

**Reply by zdz2101** [reply]
I did notice that, will remove

**Reply by bms63** [reply]
Made additional column for updating common arguments.  Hoping this helps for when new functions are created we can cross-check with common arguments.  

https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#list-of-common-arguments

**Comment by bms63** [comment]
The `derive_vars_atc` function is so specific.  The dataset_facm  has required variables that it checks for.  Maybe we should not generalize it and just leave as is?

![image](https://github.com/pharmaverse/admiral/assets/10111024/e44554b2-2931-4ddc-b2cd-e11c84fd7722)

**Reply by bundfussr** [reply]
I agree.

**Reply by bms63** [reply]
@zdz2101 @danleibovitz @manciniedoardo agree to leave as is??

**Reply by manciniedoardo** [reply]
yes agree

**Reply by zdz2101** [reply]
Looks like we'll keep dataset_facm!

**Comment by bms63** [comment]
Similarly issue for `create_single_dose_dataset` as in  `derive_var_trtemfl`, i.e. `start_date` and `end_date` are already in the function.  I think we just leave this one as is.

![image](https://github.com/pharmaverse/admiral/assets/10111024/81fe9572-56c8-479d-b562-265e4441efe2)

**Reply by zdz2101** [reply]
I agree then

**Comment by bms63** [comment]
# Check yourself before you wreck yourself

To me `check_vars` is doing something different from `check_type` and I think should be left as is.

![image](https://github.com/pharmaverse/admiral/assets/10111024/2a206a6c-9048-4951-ac65-373b1dab0f91)

## consolidate_metadata

![image](https://github.com/pharmaverse/admiral/assets/10111024/7f3519c4-05f6-439a-b13e-af88148a62da)

However,

we should bring the `check_keys` in line with the `check_type`, e.g. `filter_extreme` `check_type` is below.  Perhaps wise to check that the other functions using `check_type` follow the same - i see @inheritParams being used for a lot of them.

![image](https://github.com/pharmaverse/admiral/assets/10111024/e4d05aaf-dd9b-43d5-9ccf-2c36ba684c6b)

**Reply by bms63** [reply]
Functions with `check_type`

![image](https://github.com/pharmaverse/admiral/assets/10111024/6b32210f-94ec-405e-85c8-956fbdbb74c0)

**Reply by bundfussr** [reply]
Should we rename `check_type` and `check_keys` to `check_unique`?

**Reply by bms63** [reply]
Actually - that would make more sense to rename to `check_unique`.

**Reply by zdz2101** [reply]
I like this idea

**Reply by manciniedoardo** [reply]
I second @bundfussr 's idea, both parameters are checking uniqueness so this makes a lot of sense.

**Comment by bms63** [comment]
# Use of `by_vars` across functions

Not in our list - but in a few glances at `by_vars` I see a lot of deviation.  Shall we make an attempt to unify the language of this argument?  Perhaps some of them need to be re-named as they might do something slightly different??

![image](https://github.com/pharmaverse/admiral/assets/10111024/0ea92659-f68f-4c8e-9bf7-109ac79d1069)

![image](https://github.com/pharmaverse/admiral/assets/10111024/d1a3dbb3-fa72-4161-9954-a6c659d5f153)

![image](https://github.com/pharmaverse/admiral/assets/10111024/a3814ada-81e0-46d7-b152-3023e96feb68)

![image](https://github.com/pharmaverse/admiral/assets/10111024/55ee5cd0-17da-4061-a5c9-313ccf369d1a)

![image](https://github.com/pharmaverse/admiral/assets/10111024/4adb25d1-817d-4c10-8c8b-dcbb1fda1e0b)

![image](https://github.com/pharmaverse/admiral/assets/10111024/dc07b67d-22ee-419b-bcb7-30c8cc6f5aab)

**Reply by zdz2101** [reply]
I think the trouble here is a lot of them get passed through each other too, I think we can unify the language/description of it for sure

**Reply by bms63** [reply]
derive_param_tte is a good example where `by_vars` is different from other `by_vars`

**Comment by bms63** [comment]
## Proposal for `dataset_ref` for `derive_locf_records()`

![image](https://github.com/pharmaverse/admiral/assets/10111024/557f7b5e-c7a1-4b63-a710-d9cb3a689509)

I'm okay with replaceing this with `dataset_ref`, but this is really a lookup table similar to `derive_vars_merged_lookup()`.  We should be sure to specify in this issue.  

![image](https://github.com/pharmaverse/admiral/assets/10111024/0188888b-93ef-4bb0-91ce-81b3dbc8c2b2)

Also, similar musing as `by_vars` do we want to unifiy the messaging for each one as they differ

**Reply by bms63** [reply]
@zdz2101 @ddsjoberg @bundfussr @manciniedoardo 

apologies all - I originally had `dataset_add` , but this should be updated with `dataset_ref`, which is what @zdz2101  had originally proposed.  

Reflecting a little more: perhaps `derive_vars_merged_lookup()` argument `dataset_add` should be changed to `dataset_ref` to match proposed update and keep in line with other functions that expect lookup tables or tables with all combinations of variables.

**Reply by manciniedoardo** [reply]
I would be in favour of always using `dataset_ref` for lookup table datasets which include all combinations of visits/params, and then unifying the description langauge. In the case above i agree both can become `dataset_ref`.

**Reply by bundfussr** [reply]
I would use `dataset_ref` for `derive_locf_records()` but not for `derive_vars_merged_lookup()`. So far we use `dataset_ref` to ensure that all groups (subjects, parameters, visits, ...) which are in `dataset_ref` are also included in the output dataset (even if they are not in the input datasets (`dataset`, `dataset_add`)).
In `derive_vars_merged_lookup()` only parameters which are included in the input dataset are included in the output dataset. If a parameter is in `dataset_add` but not in `dataset`, it is not in the output dataset. Thus I would not use `dataset_ref`.

Furthermore, we should not use `dataset_ref` for `derive_var*()` functions at all because these never add observations.

**Comment by zdz2101** [comment]
Proposal to swap out ellipses/args and pass in list of certain class e.g. params, dthcaus_source, date_source, records_source. into the function argument "sources" , discussed in re-work on 20JUL2023

**Comment by bms63** [comment]
# `derive_vars_dy` (Updated #2027)

HI all,

Thinking on this update a little more

![image](https://github.com/pharmaverse/admiral/assets/10111024/d5b9ef4f-f6f2-4acd-8932-b56782316637)

So I use this function for `ADY`, `EDY`, ` ---DY ` derivations in my ADaMS, which is why I think it needs to stay as `reference_date` rather than updating to `start_date` as that could be a bit misleading - this makes me think that only `---ST` variables are allowed.

Perhaps the documentation could get tweaked a little bit??

![image](https://github.com/pharmaverse/admiral/assets/10111024/5a83318c-2ef9-44ba-a5df-7ee08951d0fe)

@bundfussr @zdz2101  @manciniedoardo what do you think? @ddsjoberg

**Reply by rossfarrugia** [reply]
agree with that - and its consistent with other functions like in {admiralonco} where we use `reference_date`: e.g. https://pharmaverse.github.io/admiralonco/main/reference/derive_param_bor.html

**Reply by zdz2101** [reply]
I am fine with a wording update, I originally paired this with start_date based on the description, but @rossfarrugia brings up a good point of the arguments in other packages (as I hadn't checked for the arguments there), if this flavor of `reference_date` is used elsewhere more reason to keep it as is.

**Reply by bundfussr** [reply]
I agree

**Reply by manciniedoardo** [reply]
I agree, as @bms63 suggested we could change "start" in the documentation to "reference".

**Comment by bms63** [comment]
# derive_param_computed wrapper style guide (No-Go!!)

Hi all,

I have in my notes a bullet on a style guide for `derive_param_computed` wrapper functions.  I think this might be a neat use case of how to make a wrapper function, things to consider, why we should be cautious on making them (e.g. maintenance of standards), and maybe some other tid-bits.  Shall we put this into the programming strategy as a section??

**Comment by bms63** [comment]
# traceability_vars

Looking into updating to `set_values_to`

![image](https://github.com/pharmaverse/admiral/assets/10111024/9bf991fe-5978-45b2-9858-10b414594942)

**Reply by bms63** [reply]
There is a suitable difference between `keep_source_vars` and `traceabiltiy_vars` ... think set values to...

**Reply by bms63** [reply]
@zdz2101  can you put your summation here please!!!

**Reply by zdz2101** [reply]
A little two-parter here:
1. In general, can we replace `traceability_vars` instances with `set_values_to` ?
2. Might be off-base, but am I correct in undertstanding that the subtle difference here with `traceability_vars` vs `keep_source_vars` is that `traceability_vars` expects a named list where the values are more "flexible" in the sense that it can be retained + modified vs just simply retained? So for the example above, we are setting `LALVDOM` as the string "AE". Alternatively, we could have done something like `LALVDOM = DOMAIN` which assuming its coming from SDTM.AE, the column `DOMAIN` is likely all set to "AE". Versus `keep_source_vars` only selects and retains the variables from the source_dataset, so `keep_source_vars = exprs(DOMAIN)` would join + keep the `DOMAIN` and user can rename on their own as `LALVDOM` after using something like `dplyr::rename()`
2b. I think we need to clarify use-cases if they are indeed different

**Reply by bundfussr** [reply]
1. Yes
2. Your analysis is correct. Maybe we should add examples to the description of the arguments in the programming strategy.

**Comment by bms63** [comment]
## Arguments using Y/N, NULL, True/False (Updated in #2033 )

I was reviewing the derive_var_ontrtfl(span_period=) argument. The arg takes values NULL or "Y". This must have been discussed previously, so really just curious about the design decisions, so I can adhere as a contribute more. Why does this arg take NULL and Y, instead of TRUE/FALSE? The latter feels more like native R syntax, and the former looks more like SAS. Was the decision to mimc SAS, or something else? THANKS!

**Comment by bms63** [comment]
# `derive_vars_period()` (no update)

@bundfussr @zdz2101 @ddsjoberg @manciniedoardo 

This function expressly says ADSL in the title and says it needs ADSL for `dataset`.  Should this be updated to `dataset_adsl` .  Looking over the ADaM IG guide these variables should be derived only for ADSL so I think it is okay...but I guess the datasets doesn't have to be an adsl dataset just needs the required variables???  Maybe adsl-like dataset with required required timing variables??

![image](https://github.com/pharmaverse/admiral/assets/10111024/3006fd07-9891-4e67-a7cc-91b9de873432)

**Reply by manciniedoardo** [reply]
If someone uses this function with a dataset that _isn't_ ADSL, but has all the required variables, then I would consider that an edge case. Most of the time you would expect to be adding these variables to ADSL. Thus I would favour changing the argument to `dataset_adsl`. Interested to hear others' thoughts on this.

**Reply by bundfussr** [reply]
In the programming strategy we have: "The first argument of `derive_` functions should be the input dataset and it should be named dataset."
Thus I would not change it.

**Reply by bms63** [reply]
ah really good point @bundfussr!  Then lets leave as is!

**Comment by bms63** [comment]
# `keep_source_vars`

Proposal to add this argument to other admiral functions.  Which ones need this added?  Functions that uses source_datasets would possibly need this update and would be a good place to start on this adventure..

**Reply by bundfussr** [reply]
I think we should add it to all selection functions which add new observations.

**Reply by bms63** [reply]
Slack post created https://pharmaverse.slack.com/archives/C028SJ83KM1/p1692633891115359

**Comment by bms63** [comment]
## Possible candidates for standardization of text in arguments 

![image](https://github.com/pharmaverse/admiral/assets/10111024/a5d4b54c-3ba8-4245-a998-a2e08aede61d)

**Reply by bms63** [reply]
Look into `set_values_to` or other candidates??

**Comment by bms63** [comment]
# `derive_var_dthcaus()`, `derive_var_extreme_dt()`, and `derive_var_extreme_dtm()`

All very similar functions - need to investigate if they can be consolidated.

**Reply by bms63** [reply]
@bundfussr to look in this issue of consolidation.

**Reply by rossfarrugia** [reply]
We already discussed `derive_var_dthcaus()` between the 3 of us and Thomas in the initial functions review and we agreed to keep this as it serves a distinct and common purpose for users - see below:
![image](https://github.com/pharmaverse/admiral/assets/82581364/eb6f275b-0d10-4a4b-ad6a-10849ae4430e)

Let's not re-do discussions from last round where decisions were already made or we'll never reach 1.0.0.

For the dt/dtm functions we also offer this for `derive_vars_dt()` & `derive_vars_dtm()` - I've personally always found this very useful to have both as some domains time is very important and others its not so having the option to use one function or the other was great when I was building our Roche ADaM templates.

**Reply by bundfussr** [reply]
If we were still in the same situation where we took the decision, I would agree. However, we are in the middle of a transition from a one function per variable/parameter approach to something else. Thus I think it makes sense to re-discuss. (We discussed expressions several times before we finally allowed them.)

**Reply by rossfarrugia** [reply]
I don't feel that we're in transition towards only ever having generic functions though either. we're probably more aiming for somewhere in the middle where we have some useful very generic functions that can do a lot but also its fine to have some targeted functions for specific common ADaM variable/parameters that will often be repeated and users value having a quick and easy way to derive them, e.g. DTHCAUS, BASE, ATOXGR, ONTRTFL, BMI, BSA to name just a few. nothing's changed in my mind for `derive_var_dthcaus()` - i still find it really useful and feel that taking it away would be at the detriment of our users. happy to hear any argument of those in favour of deprecating though.

**Reply by bundfussr** [reply]
I agree that we should not end up with generic functions only. Then we would have a major gap in the ADaM creation process. At the moment (and in the past) we replace specific function by calls to generic functions in the vignettes/templates. This has some advantages:
- less maintenance for testing and documentation for us
- less functions to learn for the user
- the derivations are clearer for the user

But also some disadvantages:
- Vignettes and templates are not well searchable
- They are not modular, e.g., you can share a complete template but not a single derivation or a subset of the derivations.

I would like to have an approach without these disadvantages.

I doubt that we would implement `derive_var_dthcaus()`, `derive_var_extreme_dt()`, and `derive_var_extreme_dtm()` today. For example you can use `derive_var_dthcaus()` instead of `derive_var_extreme_dt()` for deriving last known alive date. There are cases where you need to do this. If traceability variables are used and the date is not unique, an "arbitrary" observation is selected by `derive_var_extreme_dt()`. In `derive_var_dthcaus()` additional order variables can be specified to achieve uniqueness. But calling `derive_var_dthcaus()` to derive last known alive date is confusing. Therefore I would try to consolidate them. I would also ensure that the behavior and name of the consolidated function and its arguments is consistent with related functions, e.g., `derive_extreme_event()` is similar except that it derives new observations instead of new variable.

**Reply by rossfarrugia** [reply]
Has any user ever given this feedback? I get what you're saying that they do similar things under the hood, but when I wrote the ADSL template for Roche I found both functions very intuitive to use and I valued the very specific examples in each functional reference page that helped me to perform the common derivations DTHCAUS and LSTALVDT - as opposed to digging into a single more generic function that would take bit more work to customize for each specific purpose. I also find it easy to understand just by the function names that extreme_dt is first/last date and dthcaus is for picking out causes of death info. 

On your example of people needing to use `derive_var_dthcaus()` instead of `derive_var_extreme_dt()`, wouldn't a possible solution be to make `derive_var_extreme_dt()` (and dtm) more consistent here by adding the same logic: `If dates are equivalent, the first source will be kept, so the user should provide the inputs in the preferred order`?

I leave it to @bms63 and @manciniedoardo to make the call here, but I worry that your logic above could be applied to so many other places in {admiral} to remove the more specific functions that make it easy for users and instead have them do a bit more work with a generic function. It's a fine balance between what's best for developers and users, and for me if ever unsure I'd side with not changing it - as every deprecation impacts our users - and I could be wrong but I don't see them asking for these changes.

**Reply by bundfussr** [reply]
We got feedback (almost from the beginning) that it is hard to find the function and function call for a particular variable/parameter. https://github.com/pharmaverse/admiral/issues/2052 was created due to this feedback. Furthermore, also the opposite happens: users implement a new function although a (general) function with the required functionality exists. We observed this for example when reviewing admiralvaccine.
I think one of the reason for this issue is that we have both very general and very specific functions in our reference. Therefore I would separate them. The general functions should stay in the reference and the specific function calls are moved to a source which is easier to update and well searchable. Then users which need to implement a particular derivation could search the function call source first. If it does not include the derivation, they can look into the function reference to find a function which could be used. And ideally they store the new function call in the function call source.
Furthermore, in the function call source we could provide much more than we currently do via vignettes/templates and the few specific functions.

**Reply by manciniedoardo** [reply]
Apologies for the late reply here as I was away. 

I discussed this with @bms63 and we favour inaction here, i.e. we want to stick with the past decision of  **not condensing the functions, even if they have similar purposes**. This would minimise impact to end-users here and feel like if the functions are treating very common variables, there is limited risk in just keeping them both. 

On a more general note, we'd like to straddle the line between having only generic functions and having very specific functions, because the latter will make {admiral}'s barrier to entry too high and the former doesn't leverage the abstractions of ADaM enough. In this specific case, we feel that keeping the functions as they are is a satisfactory compromise.

**Comment by bundfussr** [comment]
I got an interesting question from a user this morning. He wants to add a flag to ADSL which flags subjects which received both DRUG A and DRUG B within the first eight days. He tried
```
adsl_treat <- adsl_dcut %>%
  derive_var_merged_exist_flag(
    dataset_add = ex_ext,
    by_vars = exprs(STUDYID, USUBJID),
    new_var = drugs_flag,
    filter_add = EXDOSE > 0,
    condition = (str_detect(EXTRT, "DRUGA") & str_detect(EXTRT, "DRUGB") & (VISITDY < 8)), 
    true_value = "Y",
    false_value = "N"
  )
```
This does not produce the desired result because there is no record in `ex_ext` where `EXTRT` equals both `"DRUGA"` and `"DRUGB"`. In principle we have the `derive_*_joined*()` functions for conditions considering more than one record. However, in `derive_var_joined_exist_flag()` you could specify `filter_joined = any(str_detect(EXTRT, "DRUGA")) & any(str_detect(EXTRT, "DRUGB")) & VISITDY < 8` but it does not provide the `dataset_add` argument.  `derive_vars_joined()` provides the `dataset_add` argument but it does not allow summary functions like `any()` in `filter_joined`.
An alternative to the `*joined*()` functions is allowing summary functions in the condition in `derive_vars_merged()`. Then `condition = any(str_detect(EXTRT, "DRUGA")) & any(str_detect(EXTRT, "DRUGB")) & VISITDY < 8` could be used.

I would propose that we
- make the arguments and the behavior of the `*joined*()` functions consistent and
- make the behavior of arguments expecting conditions consistent with respect to using summary functions.

**Comment by bundfussr** [comment]
Should we add a `description` field to all "source" objects like for `event()` and `event_joined()` (see https://github.com/pharmaverse/admiral/issues/1960)?

**Comment by bundfussr** [comment]
@rossfarrugia commented:
> @bundfussr do you think users might get confused that we have `event_sources.R` that runs `event()` to create `ADRS` objects, and `tte_sources.R` that runs `event_source()` to create `ADTTE` objects. I just got myself in a tangle trying to understand the difference between event_sources and event_source, and then figured it out. I don't have any alternative naming solution to suggest either though 😕

I agree that it may confuse users and that the naming could be improved.

What do you think?

**Reply by rossfarrugia** [reply]
@bundfussr i think the potentially easiest solution here is in admiralonco call it `response_sources.R` instead of `event_sources.R` in https://github.com/pharmaverse/admiralonco/pull/240. i think that makes it clear enough and more findable and means we don't need to change anything on admiral side. what do you think?

**Reply by bundfussr** [reply]
I think the .R files are not an issue because most users will never see them. The main issue is that we have `event()` and `event_source()`. From the names it is not clear what the difference is. In principle `event()` should have been named `event_source()` to be in line how we named such source objects before. However, this was not possible as the name was already been used.
I think we should review the names of the source objects to make them consistent and user-friendly.

**Reply by rossfarrugia** [reply]
fair point! i still think `response_sources.R` is a better filename even if just for helping us developers though. for the other point, how about `extreme_source()` instead of `event()`? Given this function is intended for use with `derive_extreme_event()`. would that be a quick fix that we could squeeze into this release to save people getting used to using `event()` that we know would likely change?

**Reply by bundfussr** [reply]
> `response_sources.R` is a better filename

I agree. I'll rename it.

Regarding naming the source objects, I would not do a quick fix now. The risk is too high that we need to change it again. I would do a thorough review after the release. Hopefully this results in names which are consistent, user-friendly, and do not change in the next years.

**Comment by bundfussr** [comment]
Sometimes variables providing codes/ids like `CMPNC` ("Standardized Medication Name Code") are character variables. In the admiral queries dataset we use `TERMID` for numeric terms and `TERMNAME` for character terms. This led to some confusion in the case that `CMPNC` is the source variable. I would propose to rename `TERMID` and `TERMNAME` to `TERMNUM` and `TERMCHAR` respectively. Then the purpose of the variables is clearer.

**Comment by bundfussr** [comment]
I have uploaded a document ([admiral Functions Overview](https://phuseaccount.sharepoint.com/:w:/r/sites/Pharmaverse-admiral/Shared%20Documents/Development%20Team%20Working%20Docs/admiral%20Functions%20Overview.docx?d=w64c31636c93c47dda066585be7a8c42b&csf=1&web=1&e=xlV57A)) which hopefully supports the clean-up discussion. I tried to classify the functions and remove inconsistencies in the naming and behavior. I also tried to think a bit further: What else beyond admiral is required to produce ADaMs?
My idea is to consider admiral as whole.
Please have a look and add comments and ideas. I am sure I have missed a lot.

Please note that I focused on creating a consistent and user-friendly structure of admiral only. I did not take into account decisions which were made in the past (see [thread](https://github.com/pharmaverse/admiral/discussions/2010#discussioncomment-6782210) above) because I am not sure which past decisions are discussible and which not.

@bms63 , when should we restart our clean-up discussions?

**Comment by bms63** [comment]
@zdz2101 Do you mind reviewing this thread for any missed actions and making them into an issue.  If things are warranting more discussion, then lets tack on to the end of @bundfussr word document so fewer places to keep track of things.

**Reply by bms63** [reply]
@zdz2101 did we finish this task?

**Reply by zdz2101** [reply]
@bms63 In terms of migrating things that were here onto the doc, yes that task is done. As for turning outstanding things on the doc into github issues/posts for 2.0 I think we might need to review it together
